src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    /// <summary>
+    /// Pipe implementation of a reader for HTTP Multipart Content Type
+    /// </summary>
+    public class MultipartPipeReader
+    {
+        private const int StackAllocThreshold = 128;
+
+        private readonly PipeReader _pipeReader;
+        private readonly MultipartBoundary _boundary;
+        private MultipartSectionPipeReader? _currentSectionReader;
+
+        private static ReadOnlySpan<byte> ColonDelimiter => new byte[] { (byte)':' };
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        /// <param name="boundary">the Multipart boundary</param>
+        /// <param name="pipeReader">a Pipe Reader</param>
+        public MultipartPipeReader(string boundary, PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = MultipartReader.DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = MultipartReader.DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional size limit for each section body.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+        /// <summary>
+        /// Read the next Multipart Section
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        public async Task<MultipartPipeSection?> ReadNextSectionAsync(CancellationToken cancellationToken = default)
+        {
+            if (_currentSectionReader == null) // first call.
+            {
+                // This section will drain any preamble data and remove the first boundary marker.
+                _currentSectionReader = new MultipartSectionPipeReader(_pipeReader, _boundary) { LengthLimit = HeadersLengthLimit };
```

Comment:
Idea (for future), could we get away with not having the MultipartSectionPipeReader class at all, tracking state in the MultipartPipeReader?